File: src/19_Remove_Nth_Node_From_End_of_List.c
// 19. Remove Nth Node From End of List
// https://leetcode.cn/problems/remove-nth-node-from-end-of-list/

// 0ms 100%
// 5.6 MB 67.44%

#include<stdio.h>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    struct ListNode *next;
};


struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
    struct ListNode dummy;
    dummy.next = head;
    struct ListNode *prev = &dummy;
    struct ListNode *tmp;
    int len = get_length(head);

    prev = &dummy;
    for (int i = 0; i < len-n ; i++) {
        prev = prev->next;
    }
    tmp = prev->next;
    prev->next = prev->next->next;
    free(tmp);

    return dummy.next;
}

int get_length(struct ListNode* head) {
    int length = 0;
    while (head) {
        ++length;
        head = head->next;
    }
    return length;
}

struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode* dummy = malloc(sizeof(struct ListNode));
    dummy->val = 0, dummy->next = head;
    struct ListNode* first = head;
    struct ListNode* second = dummy;
    
    for (int i = 0; i < n; ++i) {
        first = first->next;
    }
    while (first) {
        first = first->next;
        second = second->next;
    }
    second->next = second->next->next;
    struct ListNode* ans = dummy->next;
    free(dummy);
    return ans;
}
// 作者：LeetCode-Solution
// 链接：https://leetcode.cn/problems/remove-nth-node-from-end-of-list/solution/shan-chu-lian-biao-de-dao-shu-di-nge-jie-dian-b-61/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。


// class Solution {
// public:
//     int cur=0;
//     ListNode* removeNthFromEnd(ListNode* head, int n) {
//        if(!head) return NULL;
//        head->next = removeNthFromEnd(head->next,n);
//        cur++;
//        if(n==cur) return head->next;
//        return head;
//     }
// };

struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
    struct ListNode dummy;
    dummy.next = head;
    struct ListNode *prev = &dummy;
    struct ListNode *tmp;
    int len = 0;

    while (head != NULL) {
        head = head->next;
        len += 1;
    }

    prev = &dummy;
    for (int i = 0; i < len-n ; i++) {
        prev = prev->next;
    }
    tmp = prev->next;
    prev->next = prev->next->next;
    free(tmp);

    return dummy.next;
}

File: src/19_Remove_Nth_Node_From_End_of_List.rs
// 19. Remove Nth Node From End of List
// https://leetcode.cn/problems/remove-nth-node-from-end-of-list/

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
        next: None,
        val
        }
    }
}

impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {

        let mut dummy = Some(Box::new(ListNode{val: 0, next: head}));
        let mut slow_p = &mut dummy;
        let mut fast_p = &slow_p.clone();

        for _ in 0..=n {
            if let Some(fast_node) = fast_p {
                fast_p = &fast_node.next;
            } else {
                return None;
            }
        }

        while fast_p.is_some() {
            fast_p = &fast_p.as_ref().unwrap().next;
            slow_p = &mut slow_p.as_mut().unwrap().next;
        }

        let remove_p = &mut slow_p.as_mut().unwrap().next;
        slow_p.as_mut().unwrap().next = remove_p.as_mut().unwrap().next.take();

        dummy.unwrap().next
    }
// 作者：wanna-growth
// 链接：https://leetcode.cn/problems/remove-nth-node-from-end-of-list/solution/19-shan-chu-lian-biao-de-dao-shu-di-n-ge-dhjt/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。
}